File: client/src/app/reservation/services/reservation.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment.development';
import { Observable } from 'rxjs';
import { ReservationsResponse } from '../models/reservations-response.model';
import { ReservationFilter } from '../models/reservation-filter.model';
import { NewReservation } from '../models/new-reservation.model';
import { Reservation } from '../models/reservation.model';
import { HttpParamsBuilder } from '../../util/http-params-builder';

@Injectable({ providedIn: 'root' })
export class ReservationService {
  private readonly URL = `${environment.serverUrl}/reservations`;

  constructor(private http: HttpClient) {}

  public getReservations(
    page: number,
    size: number,
    reservationFilter: ReservationFilter
  ): Observable<ReservationsResponse> {
    const params = new HttpParamsBuilder(page, size, reservationFilter).build();
    return this.http.get<ReservationsResponse>(this.URL, { params: params });
  }

  public makeReservation(reservation: NewReservation) {
    return this.http.post<Reservation>(this.URL, reservation);
  }
}

File: server/src/main/java/com/moh/property_reservation/reservation/service/ReservationService.java
package com.moh.property_reservation.reservation.service;

import com.moh.property_reservation.reservation.dto.NewReservation;
import com.moh.property_reservation.reservation.dto.ReservationFilter;
import com.moh.property_reservation.reservation.dto.ReservationResponse;
import com.moh.property_reservation.reservation.dto.ReservationsResponse;
import org.springframework.data.domain.Pageable;

public interface ReservationService {
    ReservationsResponse getReservations(ReservationFilter filter, Pageable pageable);
    ReservationResponse makeReservation(NewReservation reservation);
}
